pkg/util/fxutil/provide_comp.go

```diff
@@ -0,0 +1,381 @@
+// into the fx constructor graph.
+//
+// For example, given:
+//
+//	type Provides struct {
+//	    My MyComponent
+//	}
+//	type Requires struct {
+//	    Dep MyDependency
+//	}
+//	func NewComponent(reqs Requires) Provides { ... }
+//
+// then:
+//
+//	ProvideComponentConstructor(NewComponent)
+//
+// will create these anonymous types:
+//
+//	type FxAwareProvides struct {
+//	    fx.Out
+//	    My MyComponent
+//	}
+//	type FxAwareRequires struct {
+//	    fx.In
+//	    Dep MyDependency
+//	}
+//
+// and essentially make this happen:
+//
+//	fx.Provide(func(fxr FxAwareRequires) fxp FxAwareProvides {
+//	    r := makePlainReqs(fxr)
+//	    p := NewComponent(r)
+//	    return makeFxAwareProvides(p)
+//	})
+func ProvideComponentConstructor(compCtorFunc interface{}) fx.Option {
+	inFxType, outFxType, outEmbeds, err := constructFxInAndOut(compCtorFunc)
+	if err != nil {
+		return fx.Error(err)
+	}
+	// build reflect.Type of the constructor function that will be provided to `fx.Provide`
+	funcFxType := reflect.FuncOf([]reflect.Type{inFxType}, []reflect.Type{outFxType}, false)
+
+	hs := &compDef.HookStorage{}
+
+	// wrapper that receives fx-aware requirements, converts them into regular requirements, and calls the constructor
+	wrapperForConstructor := func(fxAwareReqs reflect.Value) reflect.Value {
+		constructorFunc := reflect.ValueOf(compCtorFunc)
+		// invoke the regular constructor with the correct arguments
+		ctorArgs := makeConstructorArgs(constructorFunc, fxAwareReqs, hs)
+		plainOuts := constructorFunc.Call(ctorArgs)
+		maybeAppendFxLifecycle(ctorArgs, fxAwareReqs)
+		// calling `constructFxInAndOut` earlier ensures that outs has exactly 1 element
+		return makeFxAwareProvides(plainOuts[0], outFxType, outEmbeds)
+	}
+
+	// constructor function value that will inform fx what the Components are
+	fxAwareProviderFunc := reflect.MakeFunc(funcFxType, func(args []reflect.Value) []reflect.Value {
+		reqs := args[0]
+		res := wrapperForConstructor(reqs)
+		return []reflect.Value{res}
+	})
+
+	fn := fxAwareProviderFunc.Interface()
+	return fx.Provide(fn)
+}
+
+func makeConstructorArgs(constructorFunc, fxAwareReqs reflect.Value, hs *compDef.HookStorage) []reflect.Value {
+	// plain constructor either takes 0 or 1 arguments
+	if constructorFunc.Type().NumIn() == 0 {
+		return nil
+	}
+	// make requirements without fx.In
+	plainReqs := makePlainReqs(fxAwareReqs, hs)
+	// if constructor takes a single interface, extract it from reqs struct
+	if constructorFunc.Type().In(0).Kind() == reflect.Interface {
+		plainReqs = plainReqs.Field(0).Elem()
+	}
+	return []reflect.Value{plainReqs}
+}
+
+func constructFxInAndOut(compCtorFunc interface{}) (reflect.Type, reflect.Type, structEmbeds, error) {
+	// validate the constructor is a function with 0 or 1 argument(s) and 1 return value
+	ctorFuncVal := reflect.TypeOf(compCtorFunc)
+	if ctorFuncVal.Kind() != reflect.Func || ctorFuncVal.NumIn() > 1 || ctorFuncVal.NumOut() != 1 {
```

Comment:
If a constructor return a `Provides` and an error (ie `func NewComp() (Provides, error)`), would `NumOut` be equal to 2 ?

Reply:
Good point, I forgot about `error` handling, and that fx has special rules for it. I will make that fix.